HOTT-3514: Add CI jobs to deploy Terraform

amberstarlight opened the pull request:
### Jira link

[HOTT-3514](https://transformuk.atlassian.net/browse/HOTT-3514)

### What?

I have added/removed/altered:

- Heavily modifed the CircleCi `config.yml`:
  - Removed jobs/commands that have been ported across to the `trade-tariff` Orb
  - Added `terraform` jobs

### Why?

I am doing this because:

- The time has come where we can start to deploy this app to the new AWS environment.

### Notes

So far I've been adding `[ci skip]` to all my pushes on this branch, so as to avoid redeploying the app several times. I'll want to do a run on this before we merge to see what the pipeline looks like from a branch that isn't main.

It's likely that we actually want to make changes to this - as in, we may want to deploy over development for every PR, but I've had my eyes full of YAML for the past three days and I can no longer think about them in great detail without a break from it.

amberstarlight committed:
HOTT-3514: Add Terraform CI Jobs

willfish commented on .circleci/config.yml:
```diff
@@ -447,80 +398,74 @@ jobs:
+    executor:
+      name: tariff/ruby
     environment:
-      SENTRY_ENVIRONMENT: "staging"
```
Assuming this should remain staging?

amberstarlight replied:
Yep, looks like I did some copypasta

amberstarlight committed:
HOTT-3514: Bump service module to use latest

amberstarlight committed:
[ci skip] HOTT-3514: Add changes to Terraform

Lots of individual changes here but this now deploys out to AWS.

TODO items include:
- add missing secrets
- write IAM policy for SecretsManager secrets including KMS decrypt
- update healthchecking in main TF or bring them into the service

amberstarlight committed:
HOTT-3514: Bump ECS module, split IAM policy out

This adds the IAM policy for accessing the secret(s) and key(s) that
are needed to pull the secret environment variables. This also bumps
the service module to v1.5.0 that includes the IAM changes.

amberstarlight committed:
HOTT-3514: Prepare PR for review

[ci skip] as we are waiting on a PR for the infrastructure before we
can merge

Re-adds the commented out sections of the ci config file, and adds the
secret key secret required for the app to respond to healthcheck GET
requests.

amberstarlight committed:
HOTT-3514: Exclude healthcheck path from force_SSL

amberstarlight committed:
HOTT-3514: Generate image tag at apply time

amberstarlight committed:
HOTT-3514: Set docker_tag env var dynamically

amberstarlight committed:
HOTT-3514: Attempt workaround from orb issue

amberstarlight committed:
HOTT-3514: Set tf_var in executor

amberstarlight committed:
HOTT-3514: Use git command

amberstarlight committed:
HOTT-3514: Build a hideous workaround for vars

amberstarlight committed:
HOTT-3514: Ensure repo is cloned

amberstarlight committed:
HOTT-3514: Remove unused params

amberstarlight committed:
HOTT-3514: Supply BASIC_AUTH

amberstarlight committed:
HOTT-3514: Add NewRelic license key secret

amberstarlight committed:
HOTT-3514: Add IAM resource for new secret

willfish committed:
HOTT-3514: Adds healthcheckz

willfish committed:
HOTT-3514: Remove force_ssl - this is handled by the load balancer

amberstarlight committed:
HOTT-3514: Add CPU and memory variables

willfish commented on .circleci/config.yml:
```diff
@@ -2,98 +2,76 @@ version: 2.1
+              london.cloudapps.digital \
+              -n "$CF_APP-<< parameters.environment_key >>-dark"
+
+            cf add-network-policy "$CF_APP-<< parameters.environment_key >>-dark" \
```
Any thoughts on splitting the dark app name onto a new line?

amberstarlight replied:
A good idea! Done in https://github.com/trade-tariff/trade-tariff-frontend/pull/1515/commits/748a017e06ee67a24a7ea9968f1015a6b5beb093 :)

willfish replied:
:tada:

willfish commented on .circleci/config.yml:
```diff
@@ -447,80 +423,74 @@ jobs:
-    docker:
-      - image: cimg/ruby:3.2.2-node
+    executor:
+      name: tariff/ruby
```
Should this be a ruby-node executor?

amberstarlight replied:
So this uses the orb `ruby` executor, which [defaults](https://github.com/trade-tariff/trade-tariff-ci-orb/blob/main/src/executors/ruby.yml) to `3.2.2-node`.

willfish replied:
Oh. I guess the inverse of this is that when we only want a smaller ruby image we end up with node as well?

willfish commented on .circleci/config.yml:
```diff
@@ -447,80 +423,74 @@ jobs:
-    docker:
-      - image: cimg/ruby:3.2.2-node
+    executor:
+      name: tariff/ruby
```
Same again re ruby-node executor

amberstarlight replied:
As above, so below 🙏🏻

amberstarlight committed:
HOTT-3514: Fold cf add-network-policy command

amberstarlight committed:
HOTT-3514: Bump service module to v1.6.0

amberstarlight committed:
Merge branch 'main' into hott-3514-deploy-terraform
